### PR TITLE
fix(Time Value Validators): Propagate valueRequiredValidator to child component (DSP-1218)

### PR DESCRIPTION
**Pull request by mdelez**
resolves DSP-1218

### CODE REVIEW

**Comment by mdelez** [comment]
@tobiasschweizer FYI I formatted the input-value spec file so that's why there are 495 changes in that file. The relevant part is at the bottom.

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts [hunk 1]
```diff
@@ -211,6 +246,8 @@ export class TimeInputComponent extends _MatInputMixinBase implements ControlVal
     }
 
     _handleInput(): void {
+        this.dateFormControl.updateValueAndValidity();
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts:249]: Why is this needed here?

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts [hunk 2]
```diff
@@ -19,6 +21,21 @@ import { CalendarDate, CalendarPeriod, GregorianCalendarDate } from 'jdnconverti
...
 
+/** A valid time value must have both a date and a time, or both inputs must be null */
+export function dateTimeValidator(otherControl: FormControl): ValidatorFn {
+    return (control: AbstractControl): { [key: string]: any } | null => {
+
+        // valid if both date and time are null or have values
+        let invalid = !(control.value === null && otherControl.value === null || control.value !== null && otherControl.value !== null);
+
+        // ensure the time value does not contain an empty string
+        if (control.value === "" || otherControl.value === "") {
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts:32]: Couldn't this be written all in one expression?

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html [hunk 3]
```diff
@@ -1,11 +1,11 @@
 <div [formGroup]="form">
-    <mat-form-field class="child-input-component">
-      <dsp-jdn-datepicker [activeCalendar]="'Gregorian'">
-        <input matInput class="time-input-element date" [matDatepicker]="picker" [formControlName]="'date'" (dateChange)="_handleInput()" aria-label="Date" [placeholder]="dateLabel" readonly>
-        <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
-        <mat-datepicker #picker></mat-datepicker>
-      </dsp-jdn-datepicker>
-    </mat-form-field>
+    <dsp-jdn-datepicker [activeCalendar]="'Gregorian'">
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html:2]: Is this change of the wrapping (datepicker directive, mat form field) intentional?

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html [hunk 4]
```diff
@@ -15,4 +15,10 @@
             <span class="custom-error-message">Time should be given in precision HH:MM.</span>
         </mat-error>
     </mat-form-field>
+    <div class="date-form-error">
+        <mat-error *ngIf="((dateFormControl.hasError('validDateTimeRequired') || timeFormControl.hasError('validDateTimeRequired')) &&
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html:19]: Maybe it would be easier to understand this if the null checks come first.

**Review by tobiasschweizer — COMMENTED** [review]
I think we still have this t[imezone issue](https://dasch.myjetbrains.com/youtrack/issue/DSP-207#focus=Comments-4-382.0-0).

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html (hunk 3 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html:2]: It's intentional so that the css of the calendar icon works correctly. It fixes this https://dasch.myjetbrains.com/youtrack/issue/DSP-647. Two birds, one stone 🤷

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts (hunk 1 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts:249]: If we don't, `dateTimeValidator` is not triggered and the error message from it remains.
Screen Shot 2021-01-14 at 10 42 33

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html (hunk 3 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html:2]: Ok, great!

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts (hunk 1 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts:249]: Is this also the case in the other components like date and interval? Or is there something special about time?

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts (hunk 2 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts]: done in a210b88

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html (hunk 4 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.html]: done in a210b88

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.spec.ts [hunk 5]
```diff
@@ -132,4 +161,62 @@ describe('TimeInputComponent', () => {
...
+        expect(testHostComponent.timeInputComponent.form.valid).toBe(false);
+    });
+});
+
+describe('TimeInputComponent no value required', () => {
+    let testHostComponent: NoValueRequiredTestHostComponent;
+    let testHostFixture: ComponentFixture<NoValueRequiredTestHostComponent>;
+
+    beforeEach(async(() => {
+        TestBed.configureTestingModule({
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.spec.ts:188]: Do you need this setup again?

**Review by tobiasschweizer — APPROVED** [review]

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts (hunk 1 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.ts:245]: Nothing special about time (this time), we do it in the Date and Interval value components as well.

File: projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.spec.ts (hunk 5 above)
Comment (mdelez) [reply projects/dsp-ui/src/lib/viewer/values/time-value/time-input/time-input.component.spec.ts:188]: This is how it's done in the other complex value spec files. It could possibly be refactored at a later point but I don't think it's necessary.

**Review by flavens — APPROVED** [review]
